Added a Calendar control

sdoroff opened the pull request:
The control is a direct port of the Silverlight Calendar control. 
The original source is available here: https://github.com/MicrosoftArchive/SilverlightToolkit

sdoroff committed:
Adds a Calendar control ported from Silverlight

sdoroff committed:
Adds Unit Tests for the Calendar Control

kekekeks commented on src/Avalonia.Controls/Calendar/CalendarExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+    internal static class CalendarExtensions
+    {
+
+        private static Dictionary<IAvaloniaObject, Dictionary<AvaloniaProperty, bool>> _suspendedHandlers = new Dictionary<IAvaloniaObject, Dictionary<AvaloniaProperty, bool>>();
```
Why is it static?

kekekeks reviewed (approved):
LGTM, but change thread check to 1Dispatcher.UIThread.VerifyAccess`

grokys commented on src/Avalonia.Themes.Default/CalendarItem.xaml:
```diff
@@ -0,0 +1,209 @@
+                   Margin="0,4,0,4"
+                   HorizontalAlignment="Center"
+                   VerticalAlignment="Center"
+                   Text="{Binding}" />
```
Interesting, this binding produces ~50 binding errors saying:

````
Binding: Error in binding to "Avalonia.Controls.TextBlock"."Text": "Null value in expression '' at ''."
```

in Visual Studio's output window when the Calendar page in the ControlCatalog is selected. Looks like I need to look into why this might be happening, may be a problem with our binding system or notifications.

sdoroff replied:
I have a feeling the binding error are happening because the controls generated by this template are being added to the visual tree before their DataContext is getting set. I'll figure it out.

grokys replied:
This seems to have gone away, did you work out what was causing it?

grokys commented on src/Avalonia.Controls/Calendar/CalendarExtensions.cs:
```diff
@@ -0,0 +1,78 @@
+    internal static class CalendarExtensions
+    {
+
+        private static Dictionary<IAvaloniaObject, Dictionary<AvaloniaProperty, bool>> _suspendedHandlers = new Dictionary<IAvaloniaObject, Dictionary<AvaloniaProperty, bool>>();
```
I don't think that this (and the related extension methods in `CalendarExtension` is needed. As far as I can tell, this was to get around the fact that Silverlight didn't have coercion on `DependencyProperty`s. WPF does, and Avalonia does (see [here](https://github.com/AvaloniaUI/Avalonia/blob/bd17aeffe20046c7bc35e5012e24cb9827727ab3/src/Avalonia.Controls/Grid.cs#L22) for example) so I think rather than checking the value is valid in the `OnXPropertyChanged` handler and setting it back if it's not, the relevant `AvaloniaProperty` should be given a `Validate` method. The suspension stuff can then be deleted from `CalendarExtensions`.

Does that make any sense?

sdoroff replied:
That makes sense and should be pretty easy to implement.

grokys commented on src/Avalonia.Themes.Default/CalendarItem.xaml:
```diff
@@ -0,0 +1,209 @@
+                  <Style Selector="Button.Previous">
+                    <Setter Property="Template">
+                      <ControlTemplate>
+                        <Panel Cursor="Hand">
```
You should probably set `Background="Transparent"` on this panel because at the moment hit-testing is only picking up that the cursor is over the button when it's right over the `>` icon which is kinda hard to hit.

grokys commented on src/Avalonia.Themes.Default/CalendarItem.xaml:
```diff
@@ -0,0 +1,209 @@
+                  </Style>
+
+                  <Style Selector="Button.Previous">
+                    <Setter Property="Template">
```
It looks like styling the buttons using separate templates here comes from Silverlight, and I assume that's necessary because of their styling system, but in here it would be better to just set properties on the button rather than re-templating them I think. As they are, they're not re-templatable here like they are in Silverlight anyway.

grokys commented on src/Avalonia.Themes.Default/CalendarDayButton.xaml:
```diff
@@ -0,0 +1,116 @@
+    <Setter Property="Focusable" Value="False"/>
+    <Setter Property="Template">
+      <ControlTemplate>
+        <Panel>
```
Should probably set `Background="Transparent"` on this panel because the days are a bit hard to hit with the mouse currently. It looks like in Silverlight the Opacity is set on pointer enter/leave which would have the same effect, but here you're setting `Rectangle#Background`'s visibility instead.

grokys reviewed (changes requested):
This is great! I think there's a few things that could be improved in porting it over from Silverlight though.

Unfortunately it's also showing that we're still quite slow when opening the Calendar page on ControlCatalog... This will be a good test for performance.

sdoroff committed:
Correctly verified access from UI thread

Used Dispatcher.UIThread.VerifyAccess() instead of a
custom implementation

sdoroff committed:
Removed dependency on IsHandlerSuspended extensions

Replaced the IsHandlerSuspended and SetValueNoCallback extension
methods with a Validate function for the CalendarMode property and
a _displayDateIsChanging flag for the various DisplayDate properties

sdoroff committed:
Expanded HitTest regions of internal elements

Set Backgrounds to Transparent for various buttons in the control
templates to allow for expanded HitTest regions allowing for better
PointerOver and Clicking behavior

sdoroff committed:
Fixed binding errors caused by DayTitleTemplate

Fixed by initializing the DataContext before adding elements to the
VisualTree

sdoroff committed:
Merge branch 'master' into calendar-control

sdoroff commented:
I pushed each fix out as separate commit. For the future, is that the preferred method or would a single commit be better?

sdoroff committed:
Updated Theme template to avoid retemplating buttons

Removed the re-templating of the navigation buttons within the Calendar
header

sdoroff commented:
I'm not sure why a test (MemberSelectorTests.Should_Select_Child_Property_Value) failed during the AppVeyor build. The commits don't change anything that would affect that test.

jkoritzinsky commented:
That test is one of the ones we have that has spurious failures in CI. I've restarted the build for you.

grokys committed:
Merge branch 'master' into calendar-control

grokys reviewed (approved):
Just taking another look at the reason for the slowness here, and saw that it's creating 357 layers! Our deferred renderer currently creates a new layer for each control that has an opacity not set to 1, and a lot of controls here have that! I really need to fix this.

Other than that, this looks great! Thanks for doing this. I'm going to merge it in even with the performance problems it has and work on fixing those problems.
